### PR TITLE
Dep pair

### PR DESCRIPTION
**Pull request by fedepiz**
Initial implementation of Dependent Pairs + fixing the inference of DepArrays

### CODE REVIEW

File: src/main/scala/rise/core/types/Constraints.scala [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+    def unifySum(s: Sum, n: Nat)(implicit trace: Seq[Constraint]): Solution = {
+      // n = s --> 0 = s + (-n)
+      /** TODO(federico) I have no idea how this pivot system works for
+       *  unifying natural numbers. I need to solve a case like
+       *  x + 1 = y + 1. The bruteForceSum function does exactly that:
+       *  see if the pattern matches for some constant, and then unify
+       *  x and y.
+       *
+       *  Upon understanding the pivoting system, this should be cleaned up.
+       */
```
Comment (Bastacyclop) [inline src/main/scala/rise/core/types/Constraints.scala]: Is generic pivot strategy not able to unify `x + 1 = y + 1`? We can discuss the details if you want.

File: src/main/scala/rise/core/types/NatCollection.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  override def visitAndRebuild(f: Nat => Nat): Nat =
+    new NatCollectionIndexing(collection, idxs.map(idx => f(idx)))
+
+  override def substitute(subs: scala.collection.Map[Nat, Nat]): Option[Nat] = {
+    Some(visitAndRebuild(x => subs.getOrElse(x, x.substitute(subs).getOrElse(x))))
+  }
+
+}
+
+sealed abstract class NatCollection {
```
Comment (Bastacyclop) [inline src/main/scala/rise/core/types/NatCollection.scala:25]: what is a nat collection? maybe a comment would be useful here.

File: src/main/scala/rise/core/types/Type.scala [hunk 3]
```diff
@@ -112,7 +138,7 @@ final case class ArrayType(size: Nat, elemType: DataType) extends DataType {
 }
 
 final case class DepArrayType(size: Nat, fdt: NatToData) extends DataType {
-  override def toString: String = s"$size.$fdt"
+  override def toString: String = s"$size..$fdt"
```
Comment (Bastacyclop) [inline src/main/scala/rise/core/types/Type.scala:141]: I am not convinced by the `n..fdt` syntax, it looks very much like a number range to me like in `for x in 1..3`.

File: src/main/scala/rise/core/types/Solution.scala [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    substitute.n2nsInN2n(n2ns, n2n) match {
+      case id: NatToNatIdentifier => id
+      case lambda@NatToNatLambda(x, body) =>
+        val xSub = apply(x) match {
+          case n: NatIdentifier => n
+          case other => throw NonIdentifierInBinderException(lambda, other)
+        }
+        NatToNatLambda(xSub, apply(body))
+    }
+  }
```
Comment (Bastacyclop) [inline src/main/scala/rise/core/types/Solution.scala:91]: do you think the changes to the type inference will affect non-dependent programs? for example can it slow down type inference for non-dependent programs?

File: src/main/scala/rise/core/types/Constraints.scala (hunk 1 above)
Comment (fedepiz) [reply src/main/scala/rise/core/types/Constraints.scala]: Back when I started coding with the dependent pairs I was getting unification problems right there. I went back to check the code snippet, and now it seems the programs no longer take that path - bruteForceSum never seems to be invoked, even though I am using the same tests.

I believe for now I am happy removing the brute forcing and relying on the general strategy - with the caveat that we might want to re-discuss this later in case the problem shows up again

File: src/main/scala/rise/core/types/NatCollection.scala (hunk 2 above)
Comment (fedepiz) [reply src/main/scala/rise/core/types/NatCollection.scala:25]: We need dependent pairs where the first element is something akin of an array of nat - something indexable in multiple dimensions that yields nats.

Originally, @michel-steuwer and were thinking of restrcting dependent pairs first elements to nats and this kind of array, which we called a NatCollection.

Recently thought, the more I work on it, the more I think we might as well have full dependent pairs where first element is any DataType and remove the concept.

In any case, we are not there quite yet, so the dichomoty exists still.

TLDR - A NatCollection is a type-level nat array (whose value is known only at runtime though) and is probably something we are going to remove.

If this is clear, I'll add a shortened version of this as a comment

File: src/main/scala/rise/core/types/Type.scala (hunk 3 above)
Comment (fedepiz) [reply src/main/scala/rise/core/types/Type.scala:141]: I am happy for alternatives, or to overload the "n.t" syntax, as it should be unambiguous based on the kind of t (Type forArray, NatToData for DepArray) Right now we have 
(t1,t2) for pairs, 
(x ** t) for dependent pairs,
n.t for array

File: src/main/scala/rise/core/types/Solution.scala (hunk 4 above)
Comment (fedepiz) [reply src/main/scala/rise/core/types/Solution.scala:91]: For the most part, no. NatToNat and NatToData only appear when DepArrays/DepPairs are involved, and so those code paths shouldn't be taken in non-dependent programs.

There is one exception to this though -  I had to add a pass in the solution of a dep constraint, that basically does the following: for a dependent function n:K -> T, traverse T, and for all type identifiers in T, replace them with the appropriate function dependent on n. So say, if you meet the dependent function n:Nat -> n.t, with t being a type variable, this will replace t with f(n), where f is a NatToData function. This is because any type variable could be really a dependent function applied, and this dependence needs to be tracked explicitly there and then - or else it will be lost in further substitutions.
(This is what caused the old 'why dependent arrays don't infer' problem we discussed like, 9 months ago)

File: src/main/scala/rise/core/types/NatCollection.scala (hunk 2 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/types/NatCollection.scala:25]: :+1:

File: src/main/scala/rise/core/types/Type.scala (hunk 3 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/types/Type.scala:141]: overloading sounds fine to me if there is no ambiguity. otherwise maybe something like `.*` to echo the dependent pair syntax.

File: src/main/scala/rise/core/types/Solution.scala (hunk 4 above)
Comment (Bastacyclop) [reply src/main/scala/rise/core/types/Solution.scala:91]: ok, we can see how the CI times will change on shine for example when things are integrated.

**Review by Bastacyclop — APPROVED** [review]
Looks fine to me now, you can also change things like the `..` syntax in another PR.
